src/libraries/Microsoft.PowerFx.Core/Public/Marshal/ObjectMarshalerProvider.cs

```diff
@@ -0,0 +1,139 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Contracts;
+using System.Reflection;
+using System.Text;
+using System.Text.Json;
+using Microsoft.PowerFx.Core;
+using Microsoft.PowerFx.Core.IR;
+using Microsoft.PowerFx.Core.Public.Types;
+using Microsoft.PowerFx.Core.Public.Values;
+
+namespace Microsoft.PowerFx.Core
+{
+    /// <summary>
+    /// Marshal .Net classes (with fields) to <see cref="RecordValue"/>. 
+    /// This supports both strong typing and lazy marshalling. 
+    /// It will return a <see cref="ObjectMarshaler"/>. 
+    /// </summary>
+    public class ObjectMarshallerProvider : ITypeMashallerProvider
+    {
+        /// <summary>
+        /// Provides a customization point to control how properties are marshalled. 
+        /// This returns null to skip the property, else return the name it should have on the power fx record.
+        /// If this is insufficient, a caller can always implement their own marshaller and return a 
+        /// a <see cref="ObjectMarshaler"/> directly. 
+        /// </summary>
+        public Func<PropertyInfo, string> PropertyMapperFunc = (propInfo) => propInfo.Name;
+
+        /// <inheritdoc/>
+        public bool TryGetMarshaller(Type type, TypeMarshallerCache cache, int maxDepth, out ITypeMarshaller marshaler)
```

Comment:
>TryGetMarshaller [](http://example.com/codeflow?start=20&length=16)

We should have tests for derived classes, including those that override property definitions. I don't think it will have any problems, but it will be good to verity.